### PR TITLE
Removed caching from resource_static

### PR DESCRIPTION
**Pull request by orangejenny**
https://dimagi-dev.atlassian.net/browse/HI-865

During deploy, resource_versions should be
1. Cleared
1. Populated with hashes for all static files, which happens during `resource_static`
1. Appended with hashes for RequireJS bundles, which happens during `build_requirejs`

https://github.com/dimagi/commcare-hq/pull/24938 updated this logic to deal with an issue where resource_versions contained only the requirejs bundles. It did so by updating the logic in `resource_static`. However, this is confounded by 1) commands during the deploy process that `rm` resource_versions.yaml and 2) caching of resource_versions.

This PR removes the resource_versions cache. The cache lasts a day, so at best it's useful when deploying one server multiple times a day without any static files changes. I'm guessing that given frequent high-priority ICDS bugs lately, india and TCL have been more likely to be deployed multiple times a day and this is why this issue has popped up more often on those servers. To further complicate recent behavior, between the release of https://github.com/dimagi/commcare-hq/pull/24938 and the release of https://github.com/dimagi/commcare-cloud/pull/3148/ and https://github.com/dimagi/commcare-cloud/pull/3186/, there was a stretch when neither of the deploy `rm`s was doing anything, then a stretch when only one of them was doing something.

I'm removing the deploy `rm` commands in https://github.com/dimagi/commcare-cloud/pull/3222 so that all logic about removing/overwriting resource_versions is contained in the `resource_static` command.

### CODE REVIEW

**Comment by orangejenny** [comment]
Marking don't merge to prevent premature merge; this and https://github.com/dimagi/commcare-cloud/pull/3222 should be deployed together.

File: corehq/apps/hqwebapp/management/commands/resource_static.py
Comment (emord) [inline corehq/apps/hqwebapp/management/commands/resource_static.py]: Is path ever used?

File: corehq/apps/hqwebapp/management/commands/resource_static.py
Comment (emord) [inline corehq/apps/hqwebapp/management/commands/resource_static.py]: Am I reading this code correctly that resource_versions.yaml was always being overwritten? I believe that `open(path, 'w')` would have always overwritten it so the removal in commcare-cloud would have no affect. I think that if we actually needed to remove it, it would have been `'a'`.

I also think this may mean there's no dependency between this and the commacre-cloud PR

File: corehq/apps/hqwebapp/management/commands/resource_static.py
Comment (emord) [inline corehq/apps/hqwebapp/management/commands/resource_static.py]: I think that `gitinfo` is no longer used here

**Review by emord — COMMENTED** [review]
I'd approve this, but I think I've proven that I have no idea how this process is working. It seems complicated and your fixes definitely make it feel simpler and more right.

File: corehq/apps/hqwebapp/management/commands/resource_static.py
Comment (orangejenny) [reply corehq/apps/hqwebapp/management/commands/resource_static.py]: Yes, it's there for tests: https://github.com/dimagi/commcare-hq/blob/1455305a9d5ad2c1da2b560df8e2590463ccee09/corehq/apps/hqwebapp/tests/test_static_resource_command.py#L44

File: corehq/apps/hqwebapp/management/commands/resource_static.py
Comment (millerdev) [reply corehq/apps/hqwebapp/management/commands/resource_static.py]: Yes, `'w'` causes it to always be overwritten, but the previous resource versions are merged in when not overwriting:

https://github.com/dimagi/commcare-hq/blob/ea01bf98f49562baa130f48505ab361b85e93730/corehq/apps/hqwebapp/management/commands/resource_static.py#L22-L26

I'm assuming that `resource_versions.yaml` exists when `get_resource_versions(path=path)` is called; if it doesn't then that just returns an empty dict. I have not analyzed the places where that file is removed. Maybe it would make sense to have that fail hard if it does not exist but it is expected to exist?

File: corehq/apps/hqwebapp/management/commands/resource_static.py
Comment (orangejenny) [reply corehq/apps/hqwebapp/management/commands/resource_static.py]: 8db3a6e489908e85a76a8353435d5058f1440ae0

**Review by millerdev — APPROVED** [review]
It's hard to understand all the moving parts here. Cautiously approving...

File: corehq/apps/hqwebapp/management/commands/resource_static.py
Comment (orangejenny) [reply corehq/apps/hqwebapp/management/commands/resource_static.py]: > I have not analyzed the places where that file is removed.

It's only removed by the deploy script, in the places affected by https://github.com/dimagi/commcare-cloud/pull/3222. My intent with this PR is to make these two management commands the **only** places that touch resource_versions.yaml.

I like the idea of erroring if resource_versions isn't found, added in 398c5e6239d393003dd60f9514abdac892066dc0  Given that I'm also removing the `rm`s, I think this is unlikely to trigger (though it would have made this bug much more apparent), but it won't hurt.

**Comment by orangejenny** [comment]
@millerdev @emord 

I updated https://github.com/dimagi/commcare-cloud/pull/3222/ after testing more and talking with Emord: 
- 725c3010f42f211bc96d2fe408eb505670e45215 removes `prime_version_static`, which used to run early in deploy to prime the cache. My earlier statement that the cache is rarely used was wrong; it looks like it's intended to make deploy faster, so every webworker/proxy doesn't need to do versioning. However, I ran the bulk of `resource_static` in a shell on a prod webworker and found it only took a few seconds. The cache priming has been around forever, so it probably used to improve performance but doesn't seem necessary now.
- b2e8ae8ebab18200a6acc3de6639254600d0b872 moves static versioning after bower install. This is an edge case: if versioning happens, then bower is updated and that updates a particular file, that file's version hash will be incorrect.

The root cause here seems to be the `rm` during deploy, before build_requirejs. When I deploy only the commcare-cloud PR, the issue is fixed. But I do think that the caching was making the bug appear only intermittently, and it also makes it a good deal harder to reason about deploy.

Happy to hear more feedback or any suggestions for testing. I've done quite a few staging deploys this morning.

**Comment by emord** [comment]
> b2e8ae8ebab18200a6acc3de6639254600d0b872 moves static versioning after bower install. This is an edge case: if versioning happens, then bower is updated and that updates a particular file, that file's version hash will be incorrect.

Do you mean that this was an edge case/bug before? And that your current PR fixes it?

**Review by emord — APPROVED** [review]
I believe that I understand this process much more and the simplifications that are being deployed are a definite improvement

**Comment by orangejenny** [comment]
> Do you mean that this was an edge case/bug before? And that your current PR fixes it?

yes

**Comment by orangejenny** [comment]
@emord Mind approving the other PR if you're on board with it?

**Comment by millerdev** [comment]
> - 725c3010f42f211bc96d2fe408eb505670e45215 removes prime_version_static...
> - b2e8ae8ebab18200a6acc3de6639254600d0b872 moves static versioning...

What happened to these commits? Did they get removed?

**Comment by orangejenny** [comment]
@millerdev They're in the commcare-cloud PR, that's probably why they didn't link. Here's https://github.com/dimagi/commcare-cloud/pull/3222/commits/725c3010f42f211bc96d2fe408eb505670e45215 and then https://github.com/dimagi/commcare-cloud/pull/3222/commits/b2e8ae8ebab18200a6acc3de6639254600d0b872

**Pull request merged**
